mypy/semanal.py

```diff
@@ -795,65 +795,69 @@ def check_function_signature(self, fdef: FuncItem) -> None:
 
     def visit_class_def(self, defn: ClassDef) -> None:
         with self.scope.class_scope(defn.info):
-            with self.analyze_class_body(defn) as should_continue:
-                if should_continue:
-                    # Analyze class body.
-                    defn.defs.accept(self)
-
-    @contextmanager
-    def analyze_class_body(self, defn: ClassDef) -> Iterator[bool]:
-        with self.tvar_scope_frame(self.tvar_scope.class_frame()):
-            is_protocol = self.detect_protocol_base(defn)
-            self.update_metaclass(defn)
-            self.clean_up_bases_and_infer_type_variables(defn)
-            self.analyze_class_keywords(defn)
-            if self.typed_dict_analyzer.analyze_typeddict_classdef(defn):
-                yield False
-                return
-            named_tuple_info = self.named_tuple_analyzer.analyze_namedtuple_classdef(defn)
-            if named_tuple_info is not None:
-                # Temporarily clear the names dict so we don't get errors about duplicate names
-                # that were already set in build_namedtuple_typeinfo.
-                nt_names = named_tuple_info.names
-                named_tuple_info.names = SymbolTable()
-
-                self.enter_class(named_tuple_info)
-
-                yield True
+            with self.tvar_scope_frame(self.tvar_scope.class_frame()):
+                self.analyze_class(defn)
+
+    def analyze_class(self, defn: ClassDef) -> None:
+        is_protocol = self.detect_protocol_base(defn)
+        self.update_metaclass(defn)
+        self.clean_up_bases_and_infer_type_variables(defn)
+        self.analyze_class_keywords(defn)
+        if self.typed_dict_analyzer.analyze_typeddict_classdef(defn):
+            return
+        if self.analyze_namedtuple_classdef(defn):
+            return
+        self.setup_class_def_analysis(defn)
+        self.analyze_base_classes(defn)
+        defn.info.is_protocol = is_protocol
+        self.analyze_metaclass(defn)
+        defn.info.runtime_protocol = False
+        for decorator in defn.decorators:
+            self.analyze_class_decorator(defn, decorator)
+        self.analyze_class_body_common(defn)
+
+    def analyze_class_body_common(self, defn: ClassDef) -> None:
+        """Parts of class body analysis that are common to all kinds of class defs."""
+        self.enter_class(defn.info)
+        defn.defs.accept(self)
+        self.calculate_abstract_status(defn.info)
+        self.setup_type_promotion(defn)
+        self.apply_class_plugin_hooks(defn)
+        self.leave_class()
+
+    def analyze_namedtuple_classdef(self, defn: ClassDef) -> bool:
+        """Analyze class-based named tuple if the NamedTuple base class is present.
+
+        Return True only if the class is a NamedTuple class.
```

Comment:
Maybe mention that this function is here instead of in `self.named_tuple_analyzer`, because named tuples can have user defined methods, and their bodies needs to be analyzed?

Reply:
I instead added a TODO comment about moving this to live with the rest of the named tuple semantic analysis code.